security: use separate CA and server certs for the SQL service (separate from the RPC service)

knz opened the pull request:
Fixes #92069.
Fixes #39326.

TLDR: this sequence of commits makes it possible to use separate *server* certs for the RPC and SQL service when running CockroachDB nodes. This is possible even when both services are served from the same TCP port.

This turned to be *way* easier than expected because the TLS config paths for RPC and SQL were already cleanly separated in the code.

We're defining support for:
- optional `sql-server.crt` (and key). When this does not exist, crdb serves SQL using `node.crt` as usual.
- optional `ca-sql.crt` (and key); note that a separate CA can be used even when this file is not present by prepending it to `sql-server.crt`.
- new CLI commands `cockroach cert create-sql-server-ca` and `cockroach cert create-sql-server`.

See the individual commits for details.

**Note: this PR is not the time where we are doing major refactors/cleanups of this package. We're following established patterns here. This makes the change easily backportable.**

knz committed:
rpc,security: clarify the purpose of client cert methods

This commit splits `GetClientTLSConfig()` into
`GetRPCClientTLSConfig()` and `GetSQLClientTLSConfig()` to clarify
their respective purpose.

This sets the scene to split their behavior at a later stage.

Release note: None

knz committed:
rpc,security: clarify the purpose of server cert methods

This commit splits `GetServerTLSConfig()` into
`GetRPCServerTLSConfig()` and `GetSQLServerTLSConfig()` to clarify
their respective purpose.

This sets the scene to split their behavior at a later stage.

Release note: None

knz committed:
pgwire,server,security: allow use of a separate SQL cert

Prior to this patch, the same server certificate was used for both the
RPC and the SQL service. This was making it difficult to decouple
the PKI for client apps from that use for cluster operations.

This commit adds the capability to (optionally) use a separate
server certificate for the SQL service. When not present, the
same cert is used as for the RPC service, for backward-compatibility.

In this commit, the same CA is used for both node.crt and
sql-server.crt. This will be enhanced in the next commit.

Release note (cli change): The new command `cockroach cert
create-sql-server` simplifies the creation of a valid server
certificate for CockroachDB's SQL service.

Release note (security update): CockroachDB is now able to use a
separate server certificate `sql-server.crt` when accepting SQL
clients, if defined. When this file does not exist, CockroachDB still
uses `node.crt` for compatibility with previous versions.

knz committed:
server,security: enable use of a separate CA for the SQL service

Prior to this patch, the same CA was used when generating `node.crt`
and `sql-server.crt`, and also bundled in the TLS config payload
delivered to SQL clients during the initial TLS handshake.

This commit introduces support for the (optional) `ca-sql.crt`.
This is used (implicitly, when present) when using `cockroach cert
create-sql-server` and automatically bundled during the TLS handshake
for SQL clients.

Note that it is possible to use a secondary SQL server CA even
without this file present, by bundling the CA bytes as prefix
to `sql-server.crt`.

Release note (security update): CockroachDB now supports an optional
secondary CA certificate that can be used to generate/sign the
(also optional) SQL server certificate.

Release note (cli change): The new command `cockroach cert
create-sql-server-ca` can be used to generate a
`ca-sql.crt`/`ca-sql.key` combination, which can
then be used as input to `cockroach cert create-sql-server`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/92126)
<!-- Reviewable:end -->

bdarnell commented:
LGTM, although the naming is starting to get confusing. `ca-sql-server.crt` is used in the *client* configuration. And if we split `ca-client.crt` into SQL and RPC versions, the SQL server port would use `ca-sql-client.crt` as its CA for accepting client connections. Maybe that's OK? I can't decide whether it's confusing or not. 

In one of the test files we have a comment that `sql-server.crt` is signed by `ca-sql-server.crt`. That's helpful and should probably be in a non-test `certnames` file.

knz reviewed (commented):
> the naming is starting to get confusing. ca-sql-server.crt is used in the client configuration.

I can see how that's confusing. I've shortened to `ca-sql.crt`, which follows the pattern of `ca-ui.crt` already. I think that's better.

> if we split ca-client.crt into SQL and RPC versions, the SQL server port would use ca-sql-client.crt as its CA for accepting client connections.

Yes, this will be the naming if/when we split them. But note how i've chosen not to go that route here just yet.

> That's helpful and should probably be in a non-test certnames file.

Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92126#-:-NHJrXUb6RJUcR_drzZA:bykq372)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @catj-cockroach)



<!-- Sent from Reviewable.io -->

knz commented:
@JeffSwenson can you assist here. This PR as stated wants to split service certs between RPC and SQL _for client apps_.

In standalone SQL servers, there's currently confusion between:
- the *client* cert used by the SQL server to authn itself to the KV layer
- the *server* cert presented by the SQL server to RPC peers
- the *server* cert presented by the SQL server to SQL clients

All these currently use a single cert (confusingly called "client tenant cert").
We want to do away with this. 
I'm modifying this PR so that the server cert for SQL is using the same pattern as established here, if `sql-server.crt` exists. Do you have further thoughts?

knz commented:
@JeffSwenson today the backend dial test in the sqlproxy package asserts that the server cert presented by the _SQL service_ (to the proxy) contains the tenant ID. I believe this is not necessary. Can you confirm my understanding? We only need the tenant ID when talking to the KV layer IIUC.

jaylim-crl commented:
> @JeffSwenson today the backend dial test in the sqlproxy package asserts that the server cert presented by the _SQL service_ (to the proxy) contains the tenant ID. I believe this is not necessary. Can you confirm my understanding? We only need the tenant ID when talking to the KV layer IIUC.

We needed that to ensure that the proxy is routing connections to the right SQL pod. In the past, there have been cases where IP addresses were reused quickly, leading to stale routing tables within the proxy that it doesn't know about. We ended up routing a request for tenant-X to tenant-Y's SQL pod. Validating the tenant ID during routing allows the proxy to refresh its cache whenever that happens.

knz commented:
Thanks. This check does not require using the tenant client cert. It merely requires including the tenant ID in the (new) SQL server cert. I'll cook up a change for that.

knz committed:
rpc,security: clarify the purpose of the tenant client cert

Release note: None

bdarnell commented:
Does this also close #39326?

knz commented:
> Does this also close https://github.com/cockroachdb/cockroach/issues/39326?

good catch. It does.

knz committed:
security: clarify the purpose of the tenant client cert

This commit is rename-only and does not change functionality.

Release note: None

knz committed:
security: clean up the SQL client URI parsing

This commit achieves two goals:

- it ensures that the authn code does not directly
  generate an error if a URI has the scheme `crdb`
  but does not have the format `crdb://tenant/.../user/...`.
  URIs with a `crdb` scheme but a different structure
  are now skipped over.
- it properly validates that the tenant ID inside the URI
  is valid, instead of producing a panic.

Release note: None

knz committed:
security: use URIs to identify tenant IDs in server certs

Release note: None

knz committed:
security: support `--tenant-scope` with `cert create-sql-server`

This commit ensures that tenant IDs specified with `--tenant-scope`
are included in the SQL server cert.

Release note: None

knz committed:
security,sqlproxyccl: allow use of SQL server cert in standalone SQL servers

This commit makes it possible to use `sql-server.crt` for secondary
tenant servers, including in unit tests.

Release note: None

knz commented:
@JeffSwenson @jaylim-crl could you review the last 6 commits of this PR? Thanks

jeffswenson reviewed (commented):
LGTM

knz committed:
security: clarify some cert purposes further

Release note: None

knz commented on pkg/security/certnames/embedded.go:
```diff
@@ -34,6 +34,6 @@ const (
 const (
-	EmbeddedTenantCACert = "ca-client-tenant.crt" // CA for client connections
-	EmbeddedTenantCAKey  = "ca-client-tenant.key" // CA for client connections
+	EmbeddedTenantKVClientCACert = "ca-client-tenant.crt" // CA for tenant connections to the KV layer
```
Good point. Clarified in comment.

knz commented:
@bdarnell @JeffSwenson if you like this PR either of you actually needs to go through the motion of using the github "review" button to approve; we now have automation that requires it.

catj-cockroach commented:
I'm taking a look now but i wanted to make some preliminary asks that @knz and I discussed in our 1:1:

* Could we use `external` rather than `sql` for the certificate filename, and have `ui` fallback to `external`? We'd then also have `ca-external` for the CA certificate (and can have `ca-ui` fallback to `ca-external`). It's extremely unlikely that DNS and IP entries in the `ui` certificate would not be identical to the `external` certificate.
* Re: @bdarnell 's comment about the naming getting confusing, I'd argue we should deprecate `ui` certificates completely and have three (four with multi-tenancy) sets of certificates:
  * Node certificates (strictly for KV/single-tenant node-to-node communication)
  * Tenant Node certificates (strictly for SQL layer node-to-node communication)
  * External certificates (SQL, RPC, HTTP, any endpoint that would be accessed from outside the cluster)
  * Client certificates (strictly for incoming connections)





<!-- Sent from Reviewable.io -->

knz commented:
> Could we use external rather than sql for the certificate filename, and have ui fallback to external? We'd then also have ca-external for the CA certificate (and can have ca-ui fallback to ca-external). It's extremely unlikely that DNS and IP entries in the ui certificate would not be identical to the external certificate.

We can do this yes, but I'd rather have this as an _additional_ PR on top of this one. I believe that we desire to _also_ provide the capability to use separate cert/CA pairs for http vs sql, which this PR achieves. The proposal you make would be a substitute/simpler mechanism we can add on top of it.

(In the end we want both - i'm just concerned about making this PR longer than it is already)

>  I'd argue we should deprecate ui certificates completely and have three (four with multi-tenancy) sets of certificates

We should deprecate them _in our recommended deployment docs_  (where we'd recommend using  just 1 external cert for all services) but I think we want to keep the _option_ to use separate certs for the various types of services.

As to whether the specific word "ui" is what we want to keep (vs e.g. "http") there's another debate, which will involve things around backward compat and whatnot, we we can argue this later.

catj-cockroach commented:
I'd argue that we don't need to implement a difference between the `sql` certificate and the `external` certificate, given that the SQL endpoint is the one that cluster operators most likely want to set the certificate for. We may have situations where the RPC certificate and/or the HTTP certificate differ from the SQL endpoint, though I suspect that will be extremely rare. Having `external` mean "the SQL, HTTP, and RPC externally facing certificate" makes our TLS stack much easier to understand (and easier to automate with LetsEncrypt or other tools). Plus we then only need one function in `cockroach cert` for generating external certificates (as customers who intend on setting up a more complicated PKI won't use `cockroach cert` anyways)





<!-- Sent from Reviewable.io -->

catj-cockroach commented:
*[`pkg/cli/cert.go` line 146 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/92126#-NHV_Hid8SkIavJHY2G4:-NHVVdxh2JpS3a2BuW8H:bz33ju7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61df6f6a8f7c15babe82672ea94fcdb2e6e8686/pkg/cli/cert.go#L146)):*
> ```Go
> // A createSQLServerCert command generates a SQL server certificate and stores it
> // in the cert directory.
> var createSQLServerCertCmd = &cobra.Command{
> ```

Rather than doing this, could I suggest we do `createExternalServerCertCmd`? We will never have a customer who intends on having separate SQL and UI endpoints using `cockroach cert` for certificate management.


<!-- Sent from Reviewable.io -->

JeffSwenson commented:
> I'd argue that we don't need to implement a difference between the `sql` certificate and the `external` certificate, given that the SQL endpoint is the one that cluster operators most likely want to set the certificate for. We may have situations where the RPC certificate and/or the HTTP certificate differ from the SQL endpoint, though I suspect that will be extremely rare. Having `external` mean "the SQL, HTTP, and RPC externally facing certificate" makes our TLS stack much easier to understand (and easier to automate with LetsEncrypt or other tools). Plus we then only need one function in `cockroach cert` for generating external certificates (as customers who intend on setting up a more complicated PKI won't use `cockroach cert` anyways)

I like this simplification. Every cluster should be configured with internal certs that describe what the component is. They may be configured with external certs that are exposed to the end user via sql/http/rpc.

> I realize this is unrelated to this PR, but @JeffSwenson is there a reason we don't use node-to-node RPC for that?

My mental model of the tenant-client cert is it is the node <-> node cert for tenant sql servers. Unfortunately the tenant-client certificate predates the crdb:// uriSans. It is its own certificate type in order to allow encoding the tenant ID in the common name. I would be in favor of adding the uriSan to node certs and deprecating the tenant client cert.

knz commented:
@JeffSwenson 

> Every cluster should be configured with internal certs that describe what the component is. 

I do not understand this sentence. Please rephrase/extend.

> They may be configured with external certs that are exposed to the end user via sql/http/rpc.

Ditto

knz commented:
I'll try to capture the thinking in a RFC before we finalize the work here. I believe there's 2-3 different alternative proposals in this thread and i'd like to find consensus on both a strategy and some tactics to get there.
